### PR TITLE
Keycloak 25.0.1

### LINKED ISSUE #1050: Keycloak 25
### Problem Statement

Keycloak 25 is currently not supported

```
com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: 
Unrecognized field "organizationsEnabled" (class org.keycloak.representations.idm.RealmRepresentation), not marked as ignorable (141 known properties: "userFederationMappers", "rememberMe", "duplicateEmailsAllowed", "adminEventsDetailsEnabled", "users", "clientOfflineSessionMaxLifespan", "webAuthnPolicyRequireResidentKey", "webAuthnPolicyPasswordlessAvoidSameAuthenticatorRegister", "components", "otpPolicyType", "accessCodeLifespanUserAction", "id", "webAuthnPolicyAttestationConveyancePreference", "enabledEventTypes", "applications", "webAuthnPolicyPasswordlessSignatureAlgorithms", "eventsListeners", "ssoSessionMaxLifespanRememberMe", "defaultDefaultClientScopes", "webAuthnPolicyPasswordlessCreateTimeout", "clientOfflineSessionIdleTimeout", "notBefore", "publicKey", "smtpServer", "clientPolicies", "resetPasswordAllowed", "webAuthnPolicyAvoidSameAuthenticatorRegister", "accessTokenLifespanForImplicitFlow", "webAuthnPolicyPasswordlessUserVerificationRequirement", "clientScopes", "internationalizationEnabled", "defaultRole", "accessTokenLifespan", "passwordCredentialGrantAllowed", "federatedUsers", "applicationScopeMappings" [truncated]])
```

### Proposed Solution

Add support for Keycloak 25

### Environment

- Keycloak Version: 25.0.0
- keycloak-config-cli Version: 5.12.0
- Java Version: 21

### Additional information

_No response_

### Acceptance Criteria

_No response_

### PR DESCRIPTION
**Pull request by jonasvoelcker**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #1050 #1051

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: src/main/java/de/adorsys/keycloak/config/repository/ExecutionFlowRepository.java [hunk 1]
```diff
@@ -86,7 +87,7 @@ public void createExecutionFlow(
         logger.trace("Create non-top-level-flow in realm '{}' and top-level-flow '{}'", realmName, topLevelFlowAlias);
 
         AuthenticationManagementResource flowsResource = authenticationFlowRepository.getFlowResources(realmName);
-        flowsResource.addExecutionFlow(topLevelFlowAlias, executionFlowData);
+        flowsResource.addExecutionFlow(topLevelFlowAlias, new HashMap<>(executionFlowData));
```
Comment (thomasdarimont) [inline src/main/java/de/adorsys/keycloak/config/repository/ExecutionFlowRepository.java:90]: Is the order of the executionFlowData relevant here? If it is, then a LinkedHashMap that preserves the order would be more suitable here.

**Comment by thomasdarimont** [comment]
It seems that some tests related to requiredActions are now failing because Keycloak 25 now requires that a required action with the given providerId actually exist:

[Keycloak 24.0.5](https://github.com/keycloak/keycloak/blob/24.0.5/services/src/main/java/org/keycloak/services/resources/admin/AuthenticationManagementResource.java#L1069) 
vs.
[Keycloak 25.0.0](https://github.com/keycloak/keycloak/blob/25.0.0/services/src/main/java/org/keycloak/services/resources/admin/AuthenticationManagementResource.java#L1117)

**Comment by thomasdarimont** [comment]
I gave this a quick spin and added the required dummy extensions to the test execution. 
See: https://github.com/thomasdarimont/keycloak-config-cli/commit/31a894f35e972af2cf292ebccf90ea1e3b35831b
With this I "only" get 40 failed tests out of 355.

There seem to be a few incompatible changes in the admin-client libraries and more strict validation on authorization resources... but that's something for another day.

File: src/main/java/de/adorsys/keycloak/config/repository/ExecutionFlowRepository.java (hunk 1 above)
Comment (jonasvoelcker) [reply src/main/java/de/adorsys/keycloak/config/repository/ExecutionFlowRepository.java:90]: Hi @thomasdarimont, as the source type is also HashMap, I suggest that the order is not important ;)

**Comment by jonasvoelcker** [comment]
Hi @thomasdarimont,

thanks a lot for that insight, next time I need to catch the BadRequestException and display the error message. Do you think, it may be possible to include the extensions in a kind that doesn't require extension of the CI-scripts?

Unfortunately I am short on time right now, I guess I won't finish the task :(

Best Regards
Jonas

**Comment by daviddavidgit** [comment]
Thanks @jonasvoelcker and @thomasdarimont for your contributions. It would be nice to finish Keycloak 25 Upgrade as quick as possible.

The current state works on a Keycloak 25 instance. However, a lot of tests are still failing. @thomasdarimont somehow your fix does not work on my machine.

Maybe you can give some insights, on what to do next?

**Comment by jonasvoelcker** [comment]
Hi @daviddavidgit,

I am pretty sure that I won't be able to finish the task in the next few days. Also I am sure that it needs some decision about where the tool aims at. Do we want to test with custom Required Actions or not? If we change the tests to standard RAs the extension idea of @thomasdarimont would not be needed.

@francis-pouatcha @st3v0rr What do you think about this?

**Comment by thomasdarimont** [comment]
I think it's fine to change the tests since "unknown" required actions are now considered invalid by Keycloak 25.

How about either excluding the individual tests for 25+ or use an existing required action that's not enabled by default?

**Comment by bohmber** [comment]
I analyzed one of the other failing tests `ImportClientsIT#shouldUpdateRealmAddAuthorization` and it's related to [Do not export ids when exporting authorization settings](https://github.com/keycloak/keycloak/commit/8fb6d43e073471ce01583760c8c8582062fad953)

**Comment by jonasvoelcker** [comment]
Hi @bohmber,

I've invested some time on that issue now. I don't see any chance we got to keep that functionality working as the ResourcesResource only has the id-method to get the ResourceResource. If we don't have the id at hand we are kind of lost.

Do you mind to open an issue in the Keycloak-Repo so we might get some alternative or a revert?

Best Regards
Jonas

**Comment by jonasvoelcker** [comment]
Hi @thomasdarimont, do you maybe know a solution for the missing id to get the ResourceResource at hand? Or could you maybe speed up things if it is an actual error inside keycloak?

**Comment by bohmber** [comment]
Issue created in keycloak https://github.com/keycloak/keycloak/issues/30519

**Comment by jonasvoelcker** [comment]
Hi @st3v0rr @francis-pouatcha, could we please review this one?

**Comment by thomasdarimont** [comment]
Thanks for picking this up again! Just noticed that the version number is still 5.x in the branch, but main is already on 6.x.

**Comment by thomasdarimont** [comment]
Just tested this locally, works great with Keycloak 25.0.1! Great work:

```
services:

  keycloak:
    image: quay.io/keycloak/keycloak:25.0.1
    environment:

      # Keycloak Admin User
      KEYCLOAK_ADMIN: admin
      KEYCLOAK_ADMIN_PASSWORD: admin

      # Feature config, see: https://www.keycloak.org/server/features
      KC_FEATURES: preview

      # Disable specific features
      #      KC_FEATURES_DISABLED: "device-flow"

      # Logging, see: https://www.keycloak.org/server/logging
      KC_LOG_LEVEL: INFO

      # External frontend hostname, see: https://www.keycloak.org/server/hostname
      KC_HOSTNAME: localhost
      KC_HTTP_PORT: "8080"
      KC_HTTP_ENABLED: "true"
      KC_HTTP_RELATIVE_PATH: "/auth"

      KC_METRICS_ENABLED: "true"
      KC_HEALTH_ENABLED: "true"

      # Log Keycloak success events to the console
      KC_SPI_EVENTS_LISTENER_JBOSS_LOGGING_SUCCESS_LEVEL: "info"
      KC_SPI_EVENTS_LISTENER_JBOSS_LOGGING_ERROR_LEVEL: "warn"

      # Additional JVM options
      JAVA_OPTS_APPEND: "--show-version"

      # Enable Remote Debugging
      DEBUG: "true"
      DEBUG_PORT: "*:18787"

    ports:
      - 8080:8080   # HTTP
      - 8443:8443   # HTTPS
      - 18787:18787 # Java Remote Debug

    command:
      - "--verbose"
      - "start-dev"

  keycloak-config:
    #image: quay.io/adorsys/keycloak-config-cli:6.0.2-22.0.4
    image: docker.io/library/keycloak-config-cli:5.12.1-25.0.1-SNAPSHOT
    environment:
      KEYCLOAK_AVAILABILITYCHECK_ENABLED: "true"
      KEYCLOAK_AVAILABILITYCHECK_TIMEOUT: "180s"
      # see: https://github.com/adorsys/keycloak-config-cli/blob/v5.0.0/CHANGELOG.md
      IMPORT_FILES_LOCATIONS: "/config/*" # IMPORT_PATH: "/config"
      IMPORT_CACHE_ENABLED: "true" # IMPORT_FORCE: "false"
      IMPORT_VAR_SUBSTITUTION_ENABLED: "true" # IMPORT_VARSUBSTITUTION: "true"
      IMPORT_VALIDATE: "true"

      # See https://github.com/adorsys/keycloak-config-cli#log-level
      #LOGGING_LEVEL_KEYCLOAK_CONFIG_CLI: "DEBUG"
      # Note: the above does not work but _KCC does
      LOGGING_LEVEL_KCC: "DEBUG"

      # Veeeeery verbose HTTP log!
      #LOGGING_LEVEL_HTTP: "DEBUG"

      #LOGGING_LEVEL_ROOT: "DEBUG"
      LOGGING_LEVEL_ROOT: "INFO"

      KEYCLOAK_URL: "http://keycloak:8080/auth"
      KEYCLOAK_USER: "admin"
      KEYCLOAK_PASSWORD: "admin"
    volumes:
      - ./realms:/config:z
```

Example realm config file:

realms/test.yml:
```
realm: test
enabled: true
displayName: "Test"
```

**Comment by thomasdarimont** [comment]
One thing that is still missing is to configure the `unmanagedAttributePolicy`.

image

**Comment by thomasdarimont** [comment]
I'm at a conference today, so next week might be better to speak via teams.

If the UPConfig is missing, we could just copy the class into the lib and use a custom rest call to set the config.
Alternatively we could just keep the existing

    private Map<String, List<Map<String, Object>>> userProfile;

And add special handling for a map entry like "unamangedAttribitePolicy":[ {"value":"ENABLED"}] -> flatten that value when rendering the json.

This would allow us to support the ne property without having to introduce the type.

The dedicated userProfile config option was introduced to make it easier for user to config user profile without using the clumsy component syntax.

**Comment by jonasvoelcker** [comment]
Hi @bohmber, I've adapted the proposal from @thomasdarimont, if you see any problem in it, I might also roll it back. But I guess we should find a permanent solution, which fits all needs then ;)

**Comment by bohmber** [comment]
Hi @jonasvoelcker, looks ok so far, thanks

**Comment by thomasdarimont** [comment]
@jonasvoelcker that works for me too! Thank you very much :)

File: pom.xml [hunk 2]
```diff
@@ -627,7 +632,7 @@
                         <version>${maven-replacer.version}</version>
                         <executions>
                             <execution>
-                                <id>replace-pre-keycloak22</id>
+                                <id>replace-pre-keycloak19</id>
```
Comment (mme-flendly) [inline pom.xml:635]: why 19?

File: src/main/java/de/adorsys/keycloak/config/repository/ClientRepository.java [hunk 3]
```diff
@@ -182,12 +182,24 @@ public void createAuthorizationResource(String realmName, String id, ResourceRep
...
+        String resourceId = getResourceId(clientResource, resource.getName());
+        clientResource.authorization().resources().resource(resourceId).update(resource);
     }
 
-    public void removeAuthorizationResource(String realmName, String id, String resourceId) {
+    public void removeAuthorizationResource(String realmName, String id, String resourceName) {
         ClientResource clientResource = getResourceById(realmName, id);
-        clientResource.authorization().resources().resource(resourceId).remove();
+        String resourceId = getResourceId(clientResource, resourceName);
+        if (resourceId != null) {
```
Comment (mme-flendly) [inline src/main/java/de/adorsys/keycloak/config/repository/ClientRepository.java:192]: No warning, error or log when the resourceId was not found?

File: src/main/java/de/adorsys/keycloak/config/repository/ClientRepository.java [hunk 4]
```diff
@@ -218,12 +242,24 @@ public void createAuthorizationPolicy(String realmName, String id, PolicyReprese
...
+        clientResource.authorization().policies().policy(policyId).update(policy);
     }
 
-    public void removeAuthorizationPolicy(String realmName, String id, String policyId) {
+    public void removeAuthorizationPolicy(String realmName, String id, String policyName) {
         ClientResource clientResource = getResourceById(realmName, id);
-        clientResource.authorization().policies().policy(policyId).remove();
+        String policyId = getPolicyId(clientResource, policyName);
+        if (policyId != null) {
+            clientResource.authorization().policies().policy(policyId).remove();
```
Comment (mme-flendly) [inline src/main/java/de/adorsys/keycloak/config/repository/ClientRepository.java:253]: Same here, no warn, error or log

File: src/test/java/de/adorsys/keycloak/config/service/ImportAuthenticationFlowsIT.java [hunk 5]
```diff
@@ -451,9 +452,14 @@ void shouldChangeBrowserFlow() throws IOException {
         AuthenticationExecutionExportRepresentation myForms2 = getExecutionFlowFromFlow(flow, "my forms 2");
         assertThat(myForms2, notNullValue());
         assertThat(myForms2.getRequirement(), is("ALTERNATIVE"));
-        assertThat(myForms2.getPriority(), is(4));
         assertThat(myForms2.isUserSetupAllowed(), is(false));
         assertThat(myForms2.isAutheticatorFlow(), is(true));
+
+        if (VersionUtil.ge(KEYCLOAK_VERSION, "25")) {
```
Comment (mme-flendly) [inline src/test/java/de/adorsys/keycloak/config/service/ImportAuthenticationFlowsIT.java:458]: Are there some changes regarding the order/priority in the flows in version 25?

File: src/test/java/de/adorsys/keycloak/config/service/ImportAuthenticationFlowsIT.java [hunk 6]
```diff
@@ -472,10 +478,15 @@ AuthenticationFlowRepresentation assertThatBrowserFlowIsUpdated(int expectedNumb
         AuthenticationExecutionExportRepresentation myForms = getExecutionFlowFromFlow(flow, "my forms");
         assertThat(myForms, notNullValue());
         assertThat(myForms.getRequirement(), is("ALTERNATIVE"));
-        assertThat(myForms.getPriority(), is(3));
         assertThat(myForms.isUserSetupAllowed(), is(false));
         assertThat(myForms.isAutheticatorFlow(), is(true));
 
+        if (VersionUtil.ge(KEYCLOAK_VERSION, "25")) {
```
Comment (mme-flendly) [inline src/test/java/de/adorsys/keycloak/config/service/ImportAuthenticationFlowsIT.java:484]: Are there some changes regarding the order/priority in the flows in version 25?

File: contrib/scripts/generate-export-realm.sh [hunk 7]
```diff
@@ -22,7 +22,7 @@ docker run -d --rm \
   "quay.io/keycloak/keycloak:${KEYCLOAK_VERSION}" \
   start-dev
 
-while ! docker exec keycloak-export bash -c '/opt/keycloak/bin/kcadm.sh config credentials --server http://$HOSTNAME:8080/auth --realm master --user $KEYCLOAK_USER --password $KEYCLOAK_PASSWORD'; do
```
[the remark this answers is not in the document]
Comment (jonasvoelcker) [reply contrib/scripts/generate-export-realm.sh:25]: Hi @mme-flendly,

the problem is, that we don't have a user in the env-vars and the Keycloak-API does not take empty values any more. This is only a helper script to generate the realm exports for the test suite as well as changing some files towards the new version so it's not part of the deliverable.

File: pom.xml (hunk 2 above)
Comment (jonasvoelcker) [reply pom.xml:635]: It is the replacer plugin within the Keycloak 19 profile. When we introduced that mechanism, we named the replacer plugins 22 and 23, someone copied but didn't change the name.

File: src/main/java/de/adorsys/keycloak/config/repository/ClientRepository.java (hunk 3 above)
Comment (jonasvoelcker) [reply src/main/java/de/adorsys/keycloak/config/repository/ClientRepository.java:192]: In my opinion idempotency requires not existing entries not to throw errors when they get deleted. I am not sure if this needs a debug log or not but it doesn't need a warn log.

File: src/main/java/de/adorsys/keycloak/config/repository/ClientRepository.java (hunk 4 above)
Comment (jonasvoelcker) [reply src/main/java/de/adorsys/keycloak/config/repository/ClientRepository.java:253]: Explanation above 😉

File: src/test/java/de/adorsys/keycloak/config/service/ImportAuthenticationFlowsIT.java (hunk 5 above)
Comment (jonasvoelcker) [reply src/test/java/de/adorsys/keycloak/config/service/ImportAuthenticationFlowsIT.java:458]: It seems like the flows get a priority behind the highest execution, which is 25. This seems more reasonable to me than just adding some kind of random numbers before Keycloak 25.

File: src/test/java/de/adorsys/keycloak/config/service/ImportAuthenticationFlowsIT.java (hunk 6 above)
Comment (jonasvoelcker) [reply src/test/java/de/adorsys/keycloak/config/service/ImportAuthenticationFlowsIT.java:484]: Explanation above 😉

**Review by Mme-adorsys — APPROVED** [review]

File: src/main/java/de/adorsys/keycloak/config/repository/ClientRepository.java (hunk 3 above)
Comment (jonasvoelcker) [reply src/main/java/de/adorsys/keycloak/config/repository/ClientRepository.java:192]: Hi @bohmber,

you are free to contribute as well. 😉

**Pull request merged**